### ISSUE TITLE
mergeWithUnion generator?

### ISSUE THREAD

**Issue by cyclingwithelephants** [argoproj/argo-cd#14560]
# Summary

Since the merge generator discards non-matched applications, it's difficult to express things like "everything in directory 1, overriding from directory 2" and it would be nice to have some kind of merge which does not discard non-matched items

# Motivation

I am trying to adopt gitops patterns for a group of clusters which are currently very disparate. As such, there are a significant amount of cluster specific overrides. Following [an example](https://github.com/alexmt/argocd-core-cluster-management) from @alexmt I have got a directory tree looking like

```bash
.
├── base
│   ├── a
│   │   └── kustomization.yaml
│   └── b
│       └── kustomization.yaml
├── groups
│   └── env-1 # similarly to the example, this specific directory overrides `base`
│       ├── a
│       │   └── kustomization.yaml
│       ├── b
│       │   └── kustomization.yaml
│       └── c
│           └── kustomization.yaml
└── singletons
    └── cluster-1 #  cluster-1 is a member of env-1 and inherits its config, with overrides
        ├── a
        │   └── kustomization.yaml
        ├── b
        │   └── kustomization.yaml
        └── d
            └── kustomization.yaml
```

Following the example above, if I used merge as per the linked example, neither applications `c` or `d` would be generated since the merge generator discards non-matched items. To work around this, I would need to generate stubs of `c` and `d` in the base directory, and a stub of `d` in the `env-1` directory.

It would be nice if there were a way to merge in such a way that effectively does a merge, and then a union with the remaining items - I'll define this as `mergeWithUnion` for this discussion. For example, `mergeWithUnion(base, env-1)` from the directory tree above would generate applications `a`, `b`, and `c` - overriding `a` and `b` from `base`, and generating `c` from scratch

Given how I've defined `mergeWithUnion`, I'd like the ability to express `mergeWithUnion(mergeWithUnion(base, env-1),cluster-1)`

This would allow environmental and cluster specific differences (where they exist) to be a bit more obvious and reduce the need for stubbing, but also means an implicit inheritance that might not be so obvious to developers contributing to the repo (and hence potentially dangerous).

I also hope that this will simplify the process of diffing over time to further homogenise clusters since it'll be immediately obvious when there are no diffs.

# Proposal

There could be a new mergeWithUnion generator, or there could be some backwards compatible option set in the existing merge generator.

If this is a poor approach to solve this problem, I'd greatly appreciate some guidance on how best to handle this.

**Comment by crenshaw-dev** [comment]
Maybe related: https://github.com/argoproj/argo-cd/issues/11952

**Comment by cyclingwithelephants** [comment]
I do think both problems could have similar solutions i.e.

```yaml
generators:
- merge:
  mode: {whatever this merge mode ends up being called}
```

**Comment by duizabojul** [comment]
I agree with this, merge generator is too limited with only one way to merge.
